File: easy-projects/easy-common/src/main/java/xyz/srclab/common/cache/MapCache.java
package xyz.srclab.common.cache;

import xyz.srclab.annotation.Nullable;

import java.time.Duration;
import java.util.Map;
import java.util.function.Supplier;

/**
 * @author sunqian
 */
final class MapCache<K, V> implements Cache<K, V> {

    private final Map<K, V> map;

    MapCache(Map<K, V> map) {
        this.map = map;
    }

    MapCache(Supplier<Map<K, V>> mapSupplier) {
        this.map = mapSupplier.get();
    }

    @Override
    public boolean contains(K key) {
        return map.containsKey(key);
    }

    @Nullable
    @Override
    public V get(K key) {
        return map.get(key);
    }

    @Nullable
    @Override
    public V get(K key, CacheLoader<? super K, ? extends V> loader) {
        try {
            return map.computeIfAbsent(key, k -> {
                @Nullable CacheValue<? extends V> cacheValue = loader.loadDetail(k);
                if (cacheValue == null) {
                    throw new LoadNullException();
                }
                return cacheValue.value();
            });
        } catch (LoadNullException e) {
            return null;
        }
    }

    @Override
    public V getOrDefault(K key, @Nullable V defaultValue) {
        return map.getOrDefault(key, defaultValue);
    }

    @Override
    public void put(K key, @Nullable V value) {
        map.put(key, value);
    }

    @Override
    public void put(K key, CacheValue<? extends V> entry) {
        put(key, entry.value());
    }

    @Override
    public void expire(K key, long seconds) {
    }

    @Override
    public void expire(K key, Duration duration) {
    }

    @Override
    public void invalidate(K key) {
        map.remove(key);
    }

    @Override
    public void invalidateAll() {
        map.clear();
    }

    private static final class LoadNullException extends RuntimeException {

        public LoadNullException() {
            super(null, null, false, false);
        }
    }
}

File: boat-common/src/main/kotlin/xyz/srclab/common/cache/MapCache.kt
package xyz.srclab.common.cache

import xyz.srclab.common.collection.MapOps.Companion.removeAll
import java.time.Duration

class MapCache<K : Any, V>(private val map: MutableMap<K, V>) : Cache<K, V> {

    override fun getOrNull(key: K): V? {
        return map[key]
    }

    override fun getOrElse(key: K, defaultValue: V): V {
        return map.getOrDefault(key, defaultValue)
    }

    override fun getOrLoad(key: K, loader: (K) -> V): V {
        return map.computeIfAbsent(key, loader)
    }

    override fun put(key: K, value: V) {
        map[key] = value
    }

    override fun put(key: K, value: V, expirySeconds: Long) {
        put(key, value)
    }

    override fun put(key: K, value: V, expiry: Duration) {
        put(key, value)
    }

    override fun putAll(entries: Map<out K, V>) {
        map.putAll(entries)
    }

    override fun putAll(entries: Map<out K, V>, expirySeconds: Long) {
        putAll(entries)
    }

    override fun putAll(entries: Map<out K, V>, expiry: Duration) {
        putAll(entries)
    }

    override fun expiry(key: K, expirySeconds: Long) {
    }

    override fun expiry(key: K, expirySeconds: Duration) {
    }

    override fun expiryAll(keys: Iterable<K>, expirySeconds: Long) {
    }

    override fun expiryAll(keys: Iterable<K>, expirySeconds: Duration) {
    }

    override fun invalidate(key: K) {
        map.remove(key)
    }

    override fun invalidateAll(keys: Iterable<K>) {
        map.removeAll(keys)
    }

    override fun invalidateAll() {
        map.clear()
    }

    override fun cleanUp() {
    }
}